rust/lance-index/src/scalar/bloomfilter.rs

```diff
@@ -0,0 +1,1769 @@
+
+        // Get the new blocks from the builder
+        let new_blocks = builder.blocks;
+
+        // Combine existing blocks with new blocks
+        let mut all_blocks = self.blocks.clone();
+        all_blocks.extend(new_blocks);
+
+        // Create a new builder with all blocks to write them out
+        let mut combined_builder = BloomFilterIndexBuilder::try_new(
+            BloomFilterIndexBuilderParams {
+                number_of_items: self.number_of_items,
+                number_of_bytes: self.number_of_bytes,
+            },
+            self.data_type.clone(),
+        )?;
+        combined_builder.blocks = all_blocks;
+
+        // Write the updated index to dest_store
+        combined_builder.write_index(dest_store).await?;
+
+        Ok(CreatedIndex {
+            index_details: prost_types::Any::from_msg(&pb::BloomFilterIndexDetails::default())
+                .unwrap(),
+            index_version: BLOOMFILTER_INDEX_VERSION,
+        })
+    }
+
+    fn update_criteria(&self) -> UpdateCriteria {
+        UpdateCriteria::only_new_data(
+            TrainingCriteria::new(TrainingOrdering::Addresses).with_row_addr(),
+        )
+    }
+}
+
+fn default_number_of_items() -> u64 {
+    *DEFAULT_NUMBER_OF_ITEMS
+}
+
+fn default_number_of_bytes() -> usize {
+    *DEFAULT_NUMBER_OF_BYTES
+}
+
+// NumberOfItems: 8192 + NumberOfBytes: 16384(16KiB) + 8 SALT values -> Probability of false positive: 0.00057(1 in 1754)
+// reference: https://hur.st/bloomfilter/?n=8192&p=&m=16KiB&k=8
+static DEFAULT_NUMBER_OF_ITEMS: LazyLock<u64> = LazyLock::new(|| {
+    std::env::var("LANCE_BLOOMFILTER_DEFAULT_NUMBER_OF_ITEMS")
+        .unwrap_or_else(|_| "8192".to_string())
+        .parse()
+        .expect("failed to parse Lance_BLOOMFILTER_DEFAULT_NUMBER_OF_ITEMS")
+});
+
+static DEFAULT_NUMBER_OF_BYTES: LazyLock<usize> = LazyLock::new(|| {
+    std::env::var("LANCE_BLOOMFILTER_DEFAULT_NUMBER_OF_BYTES")
+        // 16384 = 16KiB
+        .unwrap_or_else(|_| "16384".to_string())
+        .parse()
+        .expect("failed to parse Lance_BLOOMFILTER_DEFAULT_NUMBER_OF_BYTES")
+});
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct BloomFilterIndexBuilderParams {
+    #[serde(default = "default_number_of_items")]
+    number_of_items: u64,
+    #[serde(default = "default_number_of_bytes")]
+    number_of_bytes: usize,
+}
+
+impl Default for BloomFilterIndexBuilderParams {
+    fn default() -> Self {
+        Self {
+            number_of_items: *DEFAULT_NUMBER_OF_ITEMS,
+            number_of_bytes: *DEFAULT_NUMBER_OF_BYTES,
+        }
+    }
+}
+
+impl BloomFilterIndexBuilderParams {
+    fn new(number_of_items: u64, number_of_bytes: usize) -> Self {
+        Self {
+            number_of_items,
+            number_of_bytes,
+        }
+    }
+}
+
+pub struct BloomFilterIndexBuilder {
+    params: BloomFilterIndexBuilderParams,
+    value_type: DataType,
+    blocks: Vec<BloomFilterStatistics>,
```

Comment:
why do we call it blocks? why not just zones?

Reply:
renamed